ls1intum/Artemis
# `Programming exercises`: Generate solution entries for structural test cases

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server ~~1~~ 3 (Atlassian Suite).

### Motivation and Context


These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
For code hints for structural test cases, we want to reuse the already existing Structure Test Oracle. Or to be more specific we want to reuse its output the test.json file.

### Description

This PR introduces the generation of SolutionEntries for structural test cases.
They are created from a combination of the test.json and the actual solution Java files. For the latter, I used the same tool that the Structure Test Oracle uses: QDox.
This is necessary as the test.json does not contain all the data that we need. For example, names of parameters as well as Generics are not included. This information is extracted from the QDox objects if possible. If for some reason the class/method/etc. was not found in the QDox object, I use a fallback system (e.g. var0, var1, etc. for parameters)

Currently, these solution entries are not used any further, but in the future, they would then be grouped into code hints. There will also be the possibility to edit them in the future.
Due to the complexity of the changes in this PR I did not implement any of this yet. They will be a part of an upcoming PR. For now, the generated solution entries are simply sent to the client and logged to the developer console. This is also the reason why this is currently an admin only button.

To summarize the features of this PR:
- Generation of solution entries for class/methods/constructors/attributes structural tests from the test.json
- Extraction of additional information from the solution code using QDox
- Support for Generics
- Currently only outputted to the developer console (And saved to the database)

### Steps for Testing

Prerequisites:
- 1 Admin (artemis_test_user_12 on TS3 and artemis_test_user_1 on TS2 works)
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to the Exercise Detail Page
3. Edit the solution repository
4. Edit the test.json in the test repository
5. Open the dev console
6. Click on the button "Create structural solution entries"
    - Please note that the changes might only be applied after the solution build plan finished, as not all test cases might be present yet)
8. Check that the solution entries have been created properly according to the solution repo and test.json
    - Check classes
    - Check attributes
    - Check constructors
    - Check methods
    - Check generic support
    - And more if you want to

If you encounter an error during testing please comment with the following information:
- The relevant parts from the test.json file
- The relevant source code
- The relevant resulting solution entries

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





| File | Line | Branch |
|---|---|---|
| StructuralTestCaseService | 90 | 70 |
| ProgrammingExerciseSolutionEntryResource::createStructuralSolutionEntries | 80 | 100 |
| Package Private Structural Models | ~90 | n/a |
| Client side methods | 100 | 100 |

### Screenshots
The (admin only) button to generate the solution entries for structural test cases:
<img width="644" alt="grafik" src="https://user-images.githubusercontent.com/18218285/160869539-5c64c134-4280-47a6-aeec-72c868519f1b.png">

The output on the console:
<img width="466" alt="grafik" src="https://user-images.githubusercontent.com/18218285/160869693-28f5dbf6-328e-4341-bad8-979b3422ae34.png">
